### DISCUSSION TITLE
How do I set custom DNS servers for a proxy?

### DISCUSSION THREAD

**Discussion by ghost** [shadowsocks/shadowsocks-rust#660] [Q&A] [answered]
I'm trying to set a public DNS server such as Cloudflare on my proxy (I don't own the server or have access to server configs, I'm just the client) but whatever I try it doesn't work!

Here's my sslocal config json:
```
{
    "servers": [
      {
        "remarks": "Shadowsocks proxy",
        "flag" : "NL",
        "server": "x.x.x.x",
        "server_port": 1234,
        "password": "xxx",
        "method": "chacha20-ietf-poly1305",
        "nameserver": "1.1.1.1"
      }
    ],
    "local_address": "127.0.0.1",
    "local_port": 1081,
    "mode": "tcp_and_udp",
    "dns": "cloudflare"
}
```

I check with [browserleaks.com](https://browserleaks.com/dns) and they are not taking over!

**Comment by zonyitoo** [comment]
The `dns` field is for customizing the DNS server that `sslocal` uses for resolving servers' domain names and target domain for ACL checking.

**Reply by zonyitoo** [reply]
What do you mean not working?

**Comment by database64128** [comment]
> I don't own the server or have access to server configs

Unfortunately, what you want to accomplish is best to be done on the server side (in `ssserver`'s config). It's still possible to configure a local resolver to use your desired upstream DNS, but it's definitely not `sslocal`'s job to do that.

**Comment by laikitleung** [comment, marked as answer]
nat a virtual ip such as 255.255.255.53,then set your proxy nameserver 255.255.255.53
iptables -t nat -A OUTPUT -p udp --dport 53 -d 255.255.255.53 -j DNAT --to-destination 1.1.1.1:53
iptables -t nat -A OUTPUT -p tcp --dport 53 -d 255.255.255.53 -j DNAT --to-destination 1.1.1.1:53

**Reply by ghost** [reply]
Thanks!
